### ISSUE TITLE
Question about step in the console output

### ISSUE THREAD

**Issue by fsmosca** [glinscott/nnue-pytorch#84]
### Input data
Training pos: 500m depth 5
Validation pos: 1m depth 10

### Command line
```
python train.py --flush_logs_every_n_steps 1 --max_epochs 200 --smart-fen-skipping --batch-size 8192 --threads 4 --num-workers 4 --gpu 1 ncsf_2021-02_pos500000000_d5_train_tn20.binpack ncsf_2021-02_pos1000000_d10_val_tn10.binpack
```

### Console output
```
Epoch 1:  39%|█████████▍              | 4846/12331 [05:59<09:15, 13.47it/s, loss=0.0136, v_num=1]
```

There are a total of 12331 steps, how this step is related to training pos of 500m less skipped positions from `--smart-fen-skipping` flag?

**Comment by Sopel97** [comment]
1 epoch is ~100M positions

**Comment by fsmosca** [comment]
This is what I used to understand correct me if I am wrong. I thought the value in --batch-size here is like a mini-batch.

Disregarding the effect of --smart-fen-skipping,  

samples = 500M
iteration = samples/batch_size = 500M/8192 = ~61035

To complete 1 epoch all samples should be considered in loss calculation. The trainer should execute ~61035 iterations to complete 1 epoch. Weights will be updated after every batch_size or 8192 samples are evaluated or weights will be updated after every single iteration. 

The step is like an iteration, weights are updated after every step.

When 1 epoch is ~100M that would mean the `--smart-fen-skipping` flag had skipped around ~400M samples from the initial input of 500M.

**Comment by Sopel97** [comment]
> When 1 epoch is ~100M that would mean the --smart-fen-skipping flag had skipped around ~400M samples from the initial input of 500M.

no, why? One epoch is always 100M positions, so in your case 100M/8192 steps. Epoch is not tied in any way to the size of the dataset

**Comment by fsmosca** [comment]
couple of epoch definitions

[deep ai](https://deepai.org/machine-learning-glossary-and-terms/epoch)
[radiopaedia](https://radiopaedia.org/articles/epoch-machine-learning#:~:text=An%20epoch%20is%20a%20term,machine%20learning%20algorithm%20has%20completed.&text=If%20the%20batch%20size%20is,is%20the%20number%20of%20iterations.)

**Comment by Sopel97** [comment]
which would work terribly in our case

**Comment by fsmosca** [comment]
Nothing will be revised except the labelling of epoch counter and steps.

If current epoch is fixed at 100M and we have 500M total samples, just wait for another 4 100M to complete before marking that 1 epoch is now completed.

**Comment by Sopel97** [comment]
This is problematic because then there's no way to produce multiple checkpoints and validation loss easly during the process. People often use datasets with tens of billions of positions.
